datumaro/plugins/data_formats/arrow/arrow_dataset.py

```diff
@@ -0,0 +1,203 @@
+        if self.__table is not None:
+            return
+
+        table = None
+        for file in self._files:
+            if not os.path.exists(file):
+                raise FileNotFoundError(errno.ENOENT, os.strerror(errno.ENOENT), file)
+
+            if self._keep_in_memory:
+                stream = pa.input_stream(file)
+            else:
+                stream = pa.memory_map(file)
+            opened_stream = pa.ipc.open_stream(stream)
+            table_ = opened_stream.read_all()
+            if table:
+                table = pa.concat_tables((table, table_))
+            else:
+                table = table_
+        self._update_table(table)
+
+    def _update_table(self, table: pa.Table) -> None:
+        self.__table = table
+        self.__batches = self.__table.to_batches()
+        self.__offsets = np.cumsum([0] + [len(i) for i in self.__batches]).tolist()
+
+    def flatten(self) -> "ArrowDataset":
+        dataset = ArrowDataset(self._files, self._keep_in_memory)
+        dataset._update_table(dataset.table.flatten())
+        return dataset
+
+    def get_batches(self, offset: int, length: int) -> List[pa.RecordBatch]:
+        assert length > 0
+
+        s_idx = _binary_search(self.__offsets, offset)
+        e_idx = _binary_search(self.__offsets, offset + length - 1)
+
+        batches = self.__batches[s_idx : e_idx + 1]
+        if self.__offsets[s_idx + 1] >= offset + length:
+            batches[0] = batches[0].slice(offset - self.__offsets[s_idx], length)
+        else:
+            batches[0] = batches[0].slice(offset - self.__offsets[s_idx])
+            batches[-1] = batches[-1].slice(0, offset + length - self.__offsets[e_idx])
+
+        return batches
+
+    def slice(self, offset: int, length: int) -> List[Dict]:
+        batches = self.get_batches(offset, length)
+        items = pa.Table.from_batches(batches).to_pylist()
+        return items
+
+    @overload
+    def __getitem__(self, key: str) -> "ArrowDataset":
+        ...
+
+    @overload
+    def __getitem__(self, key: int) -> Dict:
+        ...
+
+    @overload
+    def __getitem__(self, key: Union[range, slice]) -> List[Dict]:
+        ...
+
+    def __getitem__(
+        self, key: Union[int, range, slice, str]
+    ) -> Union[Dict, List[Dict], "ArrowDataset"]:
+        if isinstance(key, str):
+            dataset = ArrowDataset(self._files, self._keep_in_memory)
+            dataset._update_table(
+                self.table.drop([name for name in self.column_names if name != key])
+            )
+            return dataset
+        elif isinstance(key, slice):
+            key = range(*key.indices(len(self)))
+
+        if isinstance(key, range):
+            if key.step == 1 and key.stop >= key.start:
+                return self.slice(key.start, key.stop - key.start)
+
+        if isinstance(key, int):
+            if key < 0:
+                key += len(self)
+            return self.slice(key, 1)[0]
+        elif isinstance(key, range):
```

Comment:
I think this `elif` branch should move to L183 for better readability.

Reply:
Fixed in https://github.com/openvinotoolkit/datumaro/pull/931/commits/8799b145fcae5765b5952ebce33b0c7e71e5c446